### PR TITLE
Make the public and private DRBG thread local

### PR DESCRIPTION
**Pull request by kroeckx**
I still need to figure out how to delete them

### CODE REVIEW

**Comment by kroeckx** [comment]
For some reason this seems to deathlock in the drbgtest at a moment nobody should be holding the lock. I see lock(), lock(), unlock(), unlock(), lock(), lock(), unlock(), unlock(), lock(), each time of the master, and the last one hangs.

**Comment by kroeckx** [comment]
Since it's a recursive lock, and only a single thread, this just doesn't make any sense.

**Comment by kroeckx** [comment]
The second lock() call actually fails

**Comment by richsalz** [comment]
Add a debug printf in the lock function that prints %x of the lock.  I bet it's mutual deadlock of two different objects.

**Comment by kroeckx** [comment]
It seems I just don't have recursive locks

**Comment by kroeckx** [comment]
So my understanding is that pthread_rwlock_t supports recursive write locking, not recursive read locking

**Comment by kroeckx** [comment]
I think it's actually safe not to lock the master when calling drbg_setup()

**Comment by kroeckx** [comment]
They are now freed too.

I did not update the documentation because #5461 is rewriting it.

**Comment by kroeckx** [comment]
Assuming this goes in, do we want 0, 1 or 2 SSL specific DRBGs?

**Comment by mspncp** [comment]
> Can anyone outline the security justification for a per-SSL DRBG?

In the [original design](https://github.com/openssl/openssl/pull/5461/files#diff-9181ac017a6177a5f2619f65c9b7a346L56), the intention behind using a per-ssl DRBG is twofold: One is related to security, the other to performance.

### Advantages of per-SSL DRBG instances

#### Isolation of DRBG states

If the DRBG of one SSL session is compromised then this will not affect the other SSL sessions. Of course this won't help if the entire process address space has been compromised, only if you have a small heartbleed-like leak or if the random output has some defects. But in the former case the compromised DRBGs will be your least problem.

#### Reducing lock contention

Since ever SSL instance lives in a dedicated thread, there is no need for locking. IMHO this aspect is almost more important than the first one. Now that this pull request introduces per-thread instances of the global DRBGs, lock contention is not an argument anymore. But DRBG state isolation is still a useful feature, so I would keep the per-ssl implementation.

### Public vs. Private DRBG

An independent approach is the distinction between public vs. private instance, which has been promoted a lot by Kurt recently. It is based on a rigid distinction depending on whether the DRBGs output will be visible in public (e.g. used as a nonce) or is used for a secret (e.g., a session key).

Following these two orthogonal approaches has somehow led to a proliferation of DRBG instances. When the global 'private' DRBG was first introduced, it was not reseeded by `RAND_add()`, which was resolved by adding a 'master' DRBG (#4328, #4402) which seeds the 'private' and 'public' DRBG. This pull request adds two instances for every thread, and finally the question was raised whether an SSL object needs a private and public instance, two.

Having two per-ssl instances appears [a little bit overengineered](https://github.com/openssl/openssl/pull/4665#issuecomment-371081360), but it would be a consequence of the public vs. private principle. Unless one considers all SSL random numbers as private (as Rich suggested), in which case we are back at one per-ssl instance. But it should be reseeded from the thread-local private instance, not from the public one, is it is currently the case.

For me, a single per-ssl instance which is seeded form the thread-local private DRBG would be fine. But we should make a clear decision, soon.

**Comment by mspncp** [comment]
> I will extend the test in a few days.

At first sight, this pull request looks good. But to be sure, I'd like to do some additional live testing in the debugger. The multithreaded test case would be a good starting point for that.

**Comment by t-j-h** [comment]
Can you explain how you think "If the DRBG of one SSL session is compromised" can happen? And how such a compromise would not have implications elsewhere? i.e. what is the threat model that sas one DRBG can be compromised and others are not? 

Separate instances need justification - and I struggle to see a context where a DRBG can be compromised in one instance and not in another. What sort of compromise scenarios are you thinking about?

Removal of a single lock in an SSL context would be more than a little premature optimisation - considering everything else that is going on the context. What sort of scenario are you considering that would justify that and what other areas have been looked at in terms of what the performance challenges are (other that this specific item)?

**Comment by mspncp** [comment]
> Can you explain how you think "If the DRBG of one SSL session is compromised" can happen? And how such a compromise would not have implications elsewhere? i.e. what is the threat model that sas one DRBG can be compromised and others are not?

Actually I am not the right person to answer this question. The per-SSL instances were introduced by Rich in #4050 (ae3947de0952), following a [suggestion made by Pauli in June 2017](https://github.com/openssl/openssl/pull/3758#issuecomment-310938562). Their arguments sounded reasonable to me and nobody has questioned them ever since.

> Separate instances need justification - and I struggle to see a context where a DRBG can be compromised in one instance and not in another. What sort of compromise scenarios are you thinking about?

Two possible scenarios i already mentioned in my previous post: 1) the possibility that an adversary might be able to exploit a vulnerability to read a limited amount of the process' memory, and 2) the possibilty that some bits of the DRBG's internal state can be deduced from the random output by an yet unknown weakness of the AES-CTR PRNG.

After all: if DRBG's were perfect, then they would only have to be seeded once and never be reseeded, right?

> Removal of a single lock in an SSL context would be more than a little premature optimisation - considering everything else that is going on the context. What sort of scenario are you considering that would justify that and what other areas have been looked at in terms of what the performance challenges are (other that this specific item)?

Sorry, I don't understand this last question. Also, I'm more familiar with libcrypto than libssl, so I call @richsalz to the rescue...

**Comment by richsalz** [comment]
One important clarification.  An SSL object doesn't necessarily live in one unique thread.  Many servers (especially those for high-performance) will have a pool of worker threads, and the SSL object will be used by whichever thread is free.  This means that seeding the per-SSL objects from a per-thread DRBG is wrong; it's a recipe for disaster.

I think the proliferation of DRBG's is heading in the wrong direction.  Or, rather, heading in too many directions at once. This is because various people have different models/views as to what should be done.

A per-thread DRBG is nice, to avoid locking contention.  A per-SSL is nice, to increase isolation across connections.  Those two seem to be in direct conflict.

I suggest that we take a step back, and perhaps those involved go off-line and come to an agreement as to what to do.

My opinion is that a master, and then two global DRBG instances is enough for this release.  Unless something major happens we're clearly not ready to do more.

**Comment by kroeckx** [comment]
I've just pushed a new commit that removes the ssl specific DRBG, and reverts all related changes

**Comment by richsalz** [comment]
Can you explain why you think per-thread is better than per-connection?

**Comment by kroeckx** [comment]
It avoids taking a global lock per connection.

**Comment by richsalz** [comment]
Thanks, I figured that.  I am in favor of this, but I'd like to know that @paulidale thinks of the security trade-off.

**Comment by mspncp** [comment]
Since it looks like the thread local instances have a good chance of being accepted, I added a function for setting the reseeding defaults (#5576), to fix the problem mentioned in https://github.com/openssl/openssl/pull/5547#pullrequestreview-102119053.

**Comment by mspncp** [comment]
Looks good at first sight. I still would like to wait for the multithreaded DRBG test and play around with it in the debugger.

**Comment by kroeckx** [comment]
So I've added a multithread test. But it seems to leak memory because the cleanup function doesn't seem to be called when the thread exists.

**Comment by kroeckx** [comment]
There was a missing call to ossl_init_thread_start() that caused the cleanup function not to be called when the thread exited.

File: test/drbgtest.c [hunk 1]
```diff
@@ -760,6 +764,80 @@ static int test_rand_reseed(void)
...
+static int test_multi_thread(void)
+{
+    thread_t t;
+
+    run_thread(&t);
+    run_multi_thread_test();
+    wait_for_thread(t);
+    return 1;
+}
+#endif
```
[the remark this answers is not in the document]
Comment (kroeckx) [reply test/drbgtest.c]: I also tried various things, like running it under helgrind which didn't report any problems.

[the remark this answers is not in the document]
Comment (kroeckx) [reply test/drbgtest.c]: I now tried it with more threads, and helgrind actually complains.

[the remark this answers is not in the document]
Comment (kroeckx) [reply test/drbgtest.c]: It reports an error at rand_drbg_new (drbg_lib.c:187), where we check the parent's strength. I'll take the lock there.

**Comment by kroeckx** [comment]
I've pushed new commits addressing your comments. I've also fixed something helgrind complains about but I'm not sure sure is a real problem.

PS: helgrind complains a lot with that multithread test on master, which isn't really what I expected.

File: test/drbgtest.c [hunk 2]
```diff
@@ -760,6 +764,80 @@ static int test_rand_reseed(void)
...
+    time_t start = time(NULL);
+    RAND_DRBG *public, *private;
+
+    public = RAND_DRBG_get0_public();
+    private = RAND_DRBG_get0_private();
+    RAND_DRBG_set_reseed_time_interval(public, 1);
+    RAND_DRBG_set_reseed_time_interval(private, 1);
+
+    do
+    {
```
[the remark this answers is not in the document]
Comment (mspncp) [reply test/drbgtest.c]: The `do`and the `{` are still not on the same line.

**Comment by mspncp** [comment]
BTW: It would be much easier to review your corrections, if you could add them as extra fixups (`git commit --fixup  <commit-id>`) instead of amending the three existing commits and force-pushing. Then at the end you can do a `git rebase -i --autosquash`.

**Comment by kroeckx** [comment]
I didn't know about --fixup and --autosquash

**Comment by mspncp** [comment]
It's really a cool feature of git, because you can add fixups for any of your previous commits and the `--autosquash` will take care that the fixups are lined up with their counterparts when you rebase interactively at the end.

**Comment by paulidale** [comment]
I not keen on removing the per ssl drbg.  I'm not for or against having two there, but I do think at least one is beneficial.  As has been mentioned, it provides isolation between connections which can only be good.  It also avoids locking because ssl objects aren't thread safe already so making the assumption that they don't need to lock seems quite reasonable.

I'm less sure about the per thread drbgs.  I originally was okay with them when compared against my suggestion of a per ssl_ctx drbg between the sessions and the globals -- the idea of the latter is that each group of connections shares a drbg seeding each connection.  Per thread drbgs maintains the separation from the globals to some degree and will guarantee to avoid locking.  I'm not sure of the benefits of either.  Globals -> ssl session seems like it should be adequate.  However, the cost to create a drbg is low, so having more doesn't hurt and might help.

I'm not seeing why having global and per ssl drbg's means an extra lock.  Have I missed something?


The threat model in my mind is when the drbg is reversible based on viewing sufficient of its output.  Being able to compromise one session is better than compromising all sessions on a thread or ssl_ctx (or heaven forbid one of the globals).  Due to how the reseeding is done, compromising the end of a drbg chain provides essentially zero information about its predecessors' state.  So, yes you figured out that session but it gives no information about any other use of randomness.  Containment is the word.

The reputed attack mode against dual ec drbg was exactly this and I've been lead to believe the number of bytes required was modest as was the brute force component.  Assuming this happens again, how can/should we make life difficult for the attacker?  A majority of our crypto group got the pleasure of fixing the dual ec fun and took an extremely conservative approach.


Against a memory dump attack, I doubt these mitigations will ultimately help.  They might make it harder and partial leaks might be stopped.  I wouldn't bet on it.

**Comment by t-j-h** [comment]
As I understand it, *all* of the issues with the DualEC-DRBG would have made absolutely zero difference if you had a per-SSL DRBG. This is why I'm trying to understand what threat this is attempting to mitigate.

If you are concerned about a memory dump attach then I think the DRBG state is the least of your worries. And separate DRBGs doesn't protect you from a memory dump concern - nothing would other than actual separate memory access controls - and that is an entirely different issue.

**Comment by mspncp** [comment]
> This is why I'm trying to understand what threat this is attempting to mitigate.

Sorry, there must have been a misunderstanding: this pull request does not address a security issue of per-ssl drbg's. Originally, it tried to reduce lock-contention by making the `<public>` and `<private>` DRBG thread-local. When it turned out that reseeding per-ssl DRBGs from thread-local ones was "a recipe for disaster" (as [Rich called it](https://github.com/openssl/openssl/pull/5547#issuecomment-371519854)), the per-ssl DRBGs were 'sacrificed'.

If it is desired to keep both 1) per-SSL DRBGs and 2) thread-local `public` and `private` DRBGs, then one could use a separate global locked DRBG instance for reseeding the per-SSL instances, or use the `<master>` DRBG for it.

**Comment by mspncp** [comment]
> I'm not seeing why having global and per ssl drbg's means an extra lock

I understood Kurt as follows: since the per-SSL is not assigned to a fixed thread (it might be served by a thread pool), its parent needs to be locked when reseeding. This might be a problem when you have a lot of SSL connections depending on the same parent DRBG.

**Comment by kroeckx** [comment]
The motivation for this PR is avoiding to take locks. With the DRBG per SSL object it means you need instantiate a new DRBG for each connection, which means you need to take a lock a lock of the parent to get entropy and possibly to get a nonce. I'm hoping this improves performance in multi-threaded applications, but I have no idea.

But it very hard and ugly to actually use that SSL specific DRBG everywhere, see #5510 and what this PR reverts. #5510 is to be able to avoid having to take a lock, because without those changes you just used the global public/private DRBG which takes a lock. Having them per thread means it's very simple and everything is lock free, except 2 instantiations per thread, and the reseeds.

An other reasons is that we're moving to beta really soon, so I would like to at least having something that works now.

**Comment by paulidale** [comment]
My understanding was that the dual ec internal state could, _allegedly_, be inferred by examining sufficient bytes of output and then the output of the drbg could be determined forwards and backwards in time.  Having a per-ssl drbg means each ssl session would have to be broken individually which isn't a big help.  However it also segregates these from other rng uses.  That I can read your ssl traffic is pretty bad but it isn't quite as bad as being able to predict your private keys as they are generated _**in addition**_.  The old one global rng would result in everything being compromised after a ssl session was.  The current model doesn't.  This could be considered a pyrrhic victory I guess.

If the attack against duel ec were stronger such that the raw entropy that it was fed was revealed, then the parent drbg could also be compromised.  I don't believe this is the case.  All entropy passes through the seeding process which mixes it up thoroughly.

I could be mistaken since I wasn't involved when this exploded.

**Comment by paulidale** [comment]
> The motivation for this PR is avoiding to take locks. With the DRBG per SSL object it means you need instantiate a new DRBG for each connection, which means you need to take a lock a lock of the parent to get entropy and possibly to get a nonce. I'm hoping this improves performance in multi-threaded applications, but I have no idea.

Thanks, this makes sense.  I'd be surprised if there wasn't a performance improvement.  I vaguely remember some profiling output indicating that the rng lock was highly contentious.

I'm not so concerned about the worry @richsalz raised, having a ssl-drbg reseeding from different thread-drbgs doesn't seem damaging.  Ideologically it might seem naughty, but a good random number is still a good random number.  The risk is a compromise from the ssl-drbg to its seed drbg and that seems unlikely.

Unless I'm missing something ??

**Comment by mspncp** [comment]
> ...having a ssl-drbg reseeding from different thread-drbgs doesn't seem damaging...

Maybe not for the per-ssl drbg itself, but for the per-thread drbgs. I recall @kaduk reporting problems with concurrent reseeding, see commit message of 2139145b72d084a3f974a94accd7d9812de286e4.

**Comment by paulidale** [comment]
I see, the ssl drbg will have a parent drbg from the thread it was created on which needn't be the current thread.  Could we add a fake drbg that passes everything off to the current thread drbg?  Well something like that...

I wonder how this would interact with the SP800-90x standards?  It wouldn't be a chain in the strict sense.

**Comment by richsalz** [comment]
I repeat my concern that trying to mix per-object and per-thread is going to be very difficult to get right.

**Comment by paulidale** [comment]
@richsalz I'm inclined to agree.  A _use the current thread's drbg_ drbg is a bit of a kludge.  It _should_ work, but it feels like one of those shoulds that cause unmitigated long term pain.

Would a per ssl_ctx drbg alleviate some on the locking?  I think it is extra complexity with insufficient benefit.

**Comment by kroeckx** [comment]
I don't see how a drbg in the ssl_ctx improves anything. An ssl_ctx does not seem to be bound to a thread, there might possible be only 1 in the application.

If nobody objects I will just merge this.

**Comment by mspncp** [comment]
> If nobody objects I will just merge this.

I reconfirm.

**Comment by t-j-h** [comment]
-1 as we need to make a firm decision about this approach so I want to take that to a vote.

**Comment by mspncp** [comment]
> I wonder how this would interact with the SP800-90x standards? It wouldn't be a chain in the strict sense.

I don't see a contradiction to the standard. The requirements of the NIST standard apply for every single reseeding, don't they? Where is it written that a DRBG has to be reseeded from the same entropy source every time? Isn't a pool of DRBGs, from which the parent for reseeding is "chosen at random" even more unpredictable than using a single DRBG?

**Comment by paulidale** [comment]
I was thinking about where a drbg chain is defined.
I agree that it shouldn't make any difference in practice, but that doesn't mean it's allowed.

**Comment by mspncp** [comment]
@paulidale I assume we will have a lot of opportunities to discuss these type of questions when it comes to the FIPS evaluation.

**Comment by mspncp** [comment]
> -1 as we need to make a firm decision about this approach so I want to take that to a vote.

Well, that's probably the best to do, since no consent was reached on this pull request.

**Comment by kroeckx** [comment]
I assume you mean consensus, not consent.

**Comment by mspncp** [comment]
I wasn't aware that there is a difference between consensus and consent (the noun). dict.cc lists [both possibilities](https://www.dict.cc/?s=Konsens) as synonyms for  'broad agreement'.

**Comment by kroeckx** [comment]
https://www.merriam-webster.com/dictionary/consent seems to say it's archaic.

**Comment by levitte** [comment]
The big difference (*) between consent and consensus, as I understand it, is that the former is an individual thing (i.e. "do you consent with this?"), while the latter is more of a group mind thing (i.e. "do we have consensus on this?").

(*) apart the the purely grammatical difference that one is a verb and the other is a noun...

**Comment by t-j-h** [comment]
Perhaps github needs a feature for side discussions on language interpretation ... 

Consent - an explicit assent or approval
Consensus - a general agreement / understanding - not neccessarily formally determined

The main difference between the two is that silence can generally mean a consensus, but it cannot mean consent. i.e. a consensus can be from no objections, but consent requires a specific acknowledgement.

And to make that general "rule" have an exception - the concept of implied consent exists - but that isn't "consent" as such and is the subject of much legal debate as to its scope and validity in a range of contexts.

So in English they are rather different.

**Comment by mspncp** [comment]
Thank you all for the enlightenment! :-)

**Comment by mspncp** [comment]
Documentation for this pull request is being added in #5629.

**Comment by mspncp** [comment]
@openssl/omc when will the vote on this pull request terminate? It's important to now how much time remains afterwards for merging #5547 and #5629 before the code-freeze.

**Comment by mspncp** [comment]
As a side note: how about adding a 'vote' label in red, similar to 'hold'?

**Comment by mspncp** [comment]
... and adding a statement to the pr thread that a vote has been started and when it's due to terminate?

**Comment by t-j-h** [comment]
This PR needs to include a summary of the outcome from the openssl-project discussion - so that we actually record the discussion. I think we are (on that thread) in a rough consensus now.

Once that writeup is added to this PR I plan to remove the -1 so this can go ahead. I don't see a need for a formal separate vote now we seem to have a consensus - but that consensus needs to be recorded in the code (as a comment).

**Comment by mspncp** [comment]
> but that consensus needs to be recorded in the code (as a comment).

Would a writeup in the commit message be equally accepted? IMHO that would be a better place than inside the code as a comment.

**Comment by t-j-h** [comment]
I think it belongs with the code (as a design note). Other views?

**Comment by richsalz** [comment]
Some kind of comment saying "per-thread DRBG's to avoid locking contention and not clear the per-SSL separation is worth it for now."

**Comment by t-j-h** [comment]
Not quite the conversation thread - there is no justification for a per-SSL DRBG in terms of the defined threat model - the arguments are more along the lines that it couldn't hurt - but without demonstrating anything concrete in the way of benefit (at all). Your comment wording would encourage people to look for a per-SSL solution - and that I don't think we should be doing - without identifying a meaningful threat that it would address. The per-thread argument is for lock contention only - not for "security as such - and I think that came through in the discussion. Without capturing that we are highly likely to go around this loop again multiple more times.

**Comment by kroeckx** [comment]
Adding text in the drbg_lib.c file is going to create merge conflicts for the documentation changes that @mspncp is working on. He's already documenting them as per thread. So I suggest we just do it in the commit message for now.

I've rebased this because there were merge conflicts, and I've only slightly changed the commit message of the commit that makes it per thread. The 2 relevant commit messages currently say:
```
commit 2bf70d71df08c3edbd473ae59f90ff7847911b52
Author: Kurt Roeckx <kurt@roeckx.be>
Date:   Thu Mar 8 22:30:28 2018 +0100

    Don't use a ssl specific DRBG anymore
    
    Since the public and private DRBG are per thread we don't need one
    per ssl object anymore. It could also try to get entropy from a DRBG
    that's really from an other thread because the SSL object moved to an
    other thread.

commit ca6a298fe2a092af6644c4140f493fb13ad62ba1
Author: Kurt Roeckx <kurt@roeckx.be>
Date:   Wed Mar 7 19:25:55 2018 +0100

    Make the public and private DRBG thread local
    
    This avoids lock contention.
```

It's not clear what @t-j-h is trying to say exactly what I should write.

**Comment by kroeckx** [comment]
I would really like to get this in the beta, since this actually changes public header files. Which probably means I have like 12 hours left before it gets frozen.

**Comment by mspncp** [comment]
... or at least state explicitely that he removes his '-1'.

**Comment by mspncp** [comment]
BTW: I wouldn't worry so much about public headers: the only change to public headers I see is the removal of two preprocessor defines from `<openssl/evp.h>`.

```
diff --git a/include/openssl/evp.h b/include/openssl/evp.h
index 8b81b12362..f4324ab457 100644
--- a/include/openssl/evp.h
+++ b/include/openssl/evp.h
@@ -346,8 +346,6 @@ int (*EVP_CIPHER_meth_get_ctrl(const EVP_CIPHER *cipher))(EVP_CIPHER_CTX *,
 # define         EVP_CTRL_SET_PIPELINE_INPUT_BUFS        0x23
 /* Set the input buffer lengths to use for a pipelined operation */
 # define         EVP_CTRL_SET_PIPELINE_INPUT_LENS        0x24
-# define         EVP_CTRL_GET_DRBG                       0x25
-# define         EVP_CTRL_SET_DRBG                       0x26

 /* Padding modes */
 #define EVP_PADDING_PKCS7       1
```

All other changes to header files are internal.
 
 ```
 crypto/evp/evp_locl.h                  |  1 -
 crypto/include/internal/cryptlib_int.h |  2 +
 crypto/include/internal/rand_int.h     |  1 +
 include/openssl/evp.h                  |  2 -
 ssl/ssl_locl.h                         |  2 -
```

However, it _is_ a feature, so it should go in before the freeze.

**Comment by paulidale** [comment]
I think it would be better for those symbols to not get into the public headers at all.

**Comment by mspncp** [comment]
ping @t-j-h, do we get green light from you to merge this pr before the freeze?

**Comment by t-j-h** [comment]
-1 removed ... I got sidetracked on other things. Thanks for the reminder @mspncp

**Comment by mspncp** [comment]
ping @kroeckx: are you online, will you merge?

**Comment by mspncp** [comment]
I merged this pull request locally, resolving a trivial conflict in `ssl/statem/statem_srvr.c`. If Kurt does not react until 2 pm UTC, I'll push  https://github.com/mspncp/openssl/commits/pr-5547-merged to master.

**Comment by mspncp** [comment]
Ok, forget my last post ;-)

**Comment by mspncp** [comment]
Merged, thanks!

**Comment by kroeckx** [comment]
When I'm at work I can not reply, so thanks for merging this.

**Comment by mspncp** [comment]
You‘re welcome!

**Comment by mattcaswell** [comment]
This PR is causing a Travis failure. The problem would only have been found with [extended tests]. The issue is that one of the builds set the environment variable "OPENSSL_TEST_RAND_ORDER". If this is set then it randomises the order that the tests are run. In this mode the rand test often fails.

**Comment by mspncp** [comment]
See #5680. Hope that helps.

**Pull request closed without merging**